// @temporalio/ai-sdk/lib/activities.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createActivities = createActivities;
const common_1 = require("@temporalio/common");
/**
 * Creates Temporal activities for AI model invocation using the provided AI SDK provider.
 * These activities allow workflows to call AI models while maintaining Temporal's
 * execution guarantees and replay safety.
 *
 * @param provider The AI SDK provider to use for model invocations
 * @param mcpClientFactories A mapping of server names to functions to create mcp clients
 * @returns An object containing the activity functions
 *
 * @experimental The AI SDK integration is an experimental feature; APIs may change without notice.
 */
function createActivities(provider, mcpClientFactories) {
    let activities = {
        async invokeModel(args) {
            const model = provider.languageModel(args.modelId);
            return await model.doGenerate(args.options);
        },
    };
    if (mcpClientFactories !== undefined) {
        Object.entries(mcpClientFactories).forEach(([name, func]) => {
            activities = {
                ...activities,
                ...activitiesForName(name, func),
            };
        });
    }
    return activities;
}
function activitiesForName(name, mcpClientFactory) {
    async function listToolsActivity(args) {
        const mcpClient = await mcpClientFactory(args.clientArgs);
        try {
            const tools = await mcpClient.tools();
            return Object.fromEntries(Object.entries(tools).map(([k, v]) => [
                k,
                {
                    description: v.description,
                    inputSchema: v.inputSchema,
                },
            ]));
        }
        finally {
            await mcpClient.close();
        }
    }
    async function callToolActivity(args) {
        const mcpClient = await mcpClientFactory(args.clientArgs);
        try {
            const tools = await mcpClient.tools();
            const tool = tools[args.name];
            if (tool === undefined) {
                throw common_1.ApplicationFailure.retryable(`Tool ${args.name} not found.`);
            }
            return tool.execute(args.args, args.options);
        }
        finally {
            await mcpClient.close();
        }
    }
    return {
        [name + '-listTools']: listToolsActivity,
        [name + '-callTool']: callToolActivity,
    };
}
