// @temporalio/plugin/lib/plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimplePlugin = void 0;
/**
 * A unified plugin that implements multiple Temporal plugin interfaces.
 * Provides a simple way to configure clients, workers, bundlers, and connections
 * with consistent parameter resolution and merging strategies.
 *
 * @experimental Plugins is an experimental feature; APIs may change without notice.
 */
class SimplePlugin {
    options;
    /** The name of the plugin */
    name;
    /**
     * Creates a new SimplePlugin instance.
     * @param options Configuration options for the plugin
     */
    constructor(options) {
        this.options = options;
        this.name = options.name;
    }
    /**
     * Configures client options by merging plugin parameters with existing options.
     * @param options The existing client options
     * @returns Modified client options with plugin configuration applied
     */
    configureClient(options) {
        return {
            ...options,
            dataConverter: resolveDataConverter(options.dataConverter, this.options.dataConverter),
            interceptors: resolveClientInterceptors(options.interceptors, this.options.clientInterceptors),
        };
    }
    /**
     * Configures worker options by merging plugin parameters with existing options.
     * Activities and nexus services are appended, while other options are replaced.
     * @param options The existing worker options
     * @returns Modified worker options with plugin configuration applied
     */
    configureWorker(options) {
        return {
            ...options,
            dataConverter: resolveDataConverter(options.dataConverter, this.options.dataConverter),
            activities: resolveAppendObjectParameter(options.activities, this.options.activities),
            nexusServices: resolveAppendParameter(options.nexusServices, this.options.nexusServices),
            workflowsPath: resolveParameter(options.workflowsPath, this.options.workflowsPath),
            workflowBundle: resolveParameter(options.workflowBundle, this.options.workflowBundle),
            interceptors: resolveWorkerInterceptors(options.interceptors, this.options.workerInterceptors),
        };
    }
    /**
     * Configures replay worker options by merging plugin parameters with existing options.
     * @param options The existing replay worker options
     * @returns Modified replay worker options with plugin configuration applied
     */
    configureReplayWorker(options) {
        return {
            ...options,
            dataConverter: resolveDataConverter(options.dataConverter, this.options.dataConverter),
            workflowsPath: resolveParameter(options.workflowsPath, this.options.workflowsPath),
            workflowBundle: resolveParameter(options.workflowBundle, this.options.workflowBundle),
            interceptors: resolveWorkerInterceptors(options.interceptors, this.options.workerInterceptors),
        };
    }
    /**
     * Runs the worker, optionally wrapping execution in a custom context.
     * @param worker The worker instance to run
     * @param next Function to continue worker execution
     * @returns Promise that resolves when worker execution completes
     */
    runWorker(worker, next) {
        if (this.options.runContext !== undefined) {
            return this.options.runContext(() => next(worker));
        }
        return next(worker);
    }
    /**
     * Configures bundler options by merging plugin parameters with existing options.
     * @param options The existing bundle options
     * @returns Modified bundle options with plugin configuration applied
     */
    configureBundler(options) {
        return {
            ...options,
            workflowsPath: resolveRequiredParameter(options.workflowsPath, this.options.workflowsPath),
        };
    }
    /**
     * Configures connection options by merging plugin parameters with existing options.
     * Special handling for function-based API keys.
     * @param options The existing connection options
     * @returns Modified connection options with plugin configuration applied
     */
    configureConnection(options) {
        const apiKey = typeof options.apiKey === 'function' ? options.apiKey : undefined;
        return {
            ...options,
            tls: resolveParameter(options.tls, this.options.tls),
            apiKey: apiKey ?? resolveParameter(options.apiKey, this.options.apiKey),
        };
    }
    /**
     * Configures native connection options by merging plugin parameters with existing options.
     * @param options The existing native connection options
     * @returns Modified native connection options with plugin configuration applied
     */
    configureNativeConnection(options) {
        return {
            ...options,
            tls: resolveParameter(options.tls, this.options.tls),
            apiKey: resolveParameter(options.apiKey, this.options.apiKey),
        };
    }
}
exports.SimplePlugin = SimplePlugin;
function resolveParameterWithResolution(existing, parameter, resolve) {
    if (parameter === undefined) {
        return existing;
    }
    if (typeof parameter === 'function') {
        // @ts-expect-error Can't infer that parameter is a function
        return parameter(existing);
    }
    if (existing === undefined) {
        return parameter;
    }
    return resolve(existing, parameter);
}
function resolveRequiredParameter(existing, parameter) {
    return resolveParameterWithResolution(existing, parameter, (_existing, param) => param);
}
function resolveParameter(existing, parameter) {
    return resolveParameterWithResolution(existing, parameter, (_existing, param) => param);
}
function resolveAppendParameter(existing, parameter) {
    if (parameter === undefined) {
        return existing;
    }
    return resolveParameterWithResolution(existing ?? [], parameter, (existing, param) => existing.concat(param));
}
function resolveAppendObjectParameter(existing, parameter) {
    if (parameter === undefined) {
        return existing;
    }
    return resolveParameterWithResolution(existing ?? {}, parameter, (existing, param) => ({ ...existing, ...param }));
}
function resolveDataConverter(existing, parameter) {
    return resolveParameterWithResolution(existing, parameter, (existing, parameter) => ({
        ...existing,
        ...parameter,
    }));
}
function resolveClientInterceptors(existing, parameter) {
    return resolveParameterWithResolution(existing, parameter, (existing, parameter) => ({
        workflow: tryConcat(modernWorkflowInterceptors(existing?.workflow), modernWorkflowInterceptors(parameter?.workflow)),
        schedule: tryConcat(existing?.schedule, parameter?.schedule),
    }));
}
function resolveWorkerInterceptors(existing, parameter) {
    return resolveParameterWithResolution(existing, parameter, (existing, parameter) => ({
        client: resolveClientInterceptors(existing.client, parameter.client),
        activity: resolveAppendParameter(existing.activity, parameter.activity),
        nexus: resolveAppendParameter(existing.nexus, parameter.nexus),
        workflowModules: resolveAppendParameter(existing.workflowModules, parameter.workflowModules),
    }));
}
function modernWorkflowInterceptors(
// eslint-disable-next-line deprecation/deprecation
interceptors) {
    if (interceptors === undefined || Array.isArray(interceptors)) {
        return interceptors;
    }
    throw new Error("Simple plugin doesn't support old style workflow client interceptors");
}
function tryConcat(left, right) {
    if (right === undefined) {
        return left;
    }
    if (left === undefined) {
        return right;
    }
    return left.concat(right);
}
