// @temporalio/envconfig/lib/envconfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadClientConfig = loadClientConfig;
exports.loadClientConfigFromToml = loadClientConfigFromToml;
exports.clientConfigToToml = clientConfigToToml;
exports.loadClientConfigProfile = loadClientConfigProfile;
exports.loadClientConnectConfig = loadClientConnectConfig;
exports.toClientOptions = toClientOptions;
exports.toTLSConfig = toTLSConfig;
const encoding_1 = require("@temporalio/common/lib/encoding");
const envconfig_toml_1 = require("./envconfig-toml");
const utils_1 = require("./utils");
function loadClientConfig(options) {
    return (0, utils_1.fromTomlConfig)((0, envconfig_toml_1.tomlLoadClientConfig)(options));
}
function loadClientConfigFromToml(tomlData, options) {
    return (0, utils_1.fromTomlConfig)((0, envconfig_toml_1.loadFromTomlData)((0, encoding_1.decode)(tomlData), options.strict));
}
function clientConfigToToml(config) {
    return (0, envconfig_toml_1.configToTomlData)((0, utils_1.toTomlConfig)(config));
}
function loadClientConfigProfile(options = {}) {
    return (0, utils_1.fromTomlProfile)((0, envconfig_toml_1.tomlLoadClientConfigProfile)(options));
}
function loadClientConnectConfig(options = {}) {
    return toClientOptions(loadClientConfigProfile(options));
}
function toClientOptions(profile) {
    // TLS is enabled if we have an explicit TLS config, or if an api key is provided.
    const tls = toTLSConfig(profile.tls) ?? (profile.apiKey !== undefined ? true : undefined);
    return {
        namespace: profile.namespace,
        connectionOptions: {
            address: profile.address,
            apiKey: profile.apiKey,
            tls,
            metadata: profile.grpcMeta,
        },
    };
}
function toTLSConfig(config) {
    if (config === undefined) {
        return undefined;
    }
    if (config.disabled === true) {
        return false;
    }
    const serverRootCACert = (0, utils_1.loadConfigData)(config.serverCACert);
    const crtBuffer = (0, utils_1.loadConfigData)(config.clientCert);
    const keyBuffer = (0, utils_1.loadConfigData)(config.clientKey);
    const tlsConfig = {
        serverNameOverride: config.serverName,
        serverRootCACertificate: serverRootCACert,
        clientCertPair: crtBuffer && keyBuffer
            ? {
                crt: crtBuffer,
                key: keyBuffer,
            }
            : undefined,
    };
    return tlsConfig;
}
