// @temporalio/envconfig/lib/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadConfigData = loadConfigData;
exports.fromTomlProfile = fromTomlProfile;
exports.toTomlProfile = toTomlProfile;
exports.fromTomlTLS = fromTomlTLS;
exports.toTomlTLS = toTomlTLS;
exports.fromTomlConfig = fromTomlConfig;
exports.toTomlConfig = toTomlConfig;
exports.toPathAndData = toPathAndData;
const fs_1 = require("fs");
const internal_workflow_1 = require("@temporalio/common/lib/internal-workflow");
const encoding_1 = require("@temporalio/common/lib/encoding");
const envconfig_toml_1 = require("./envconfig-toml");
/**
 * Loads configuration data from a {@link ConfigDataSource} and returns it as a Uint8Array.
 *
 * @experimental Environment configuration is new feature and subject to change.
 */
function loadConfigData(source) {
    if (!source)
        return undefined;
    if ('path' in source) {
        return Uint8Array.from((0, fs_1.readFileSync)(source.path));
    }
    return typeof source.data === 'string' ? (0, encoding_1.encode)(source.data) : source.data;
}
/**
 * Converts a TOML profile structure to a {@link ClientConfigProfile}.
 *
 * @experimental Environment configuration is new feature and subject to change.
 */
function fromTomlProfile(tomlProfile) {
    let grpcMeta = undefined;
    if (tomlProfile.grpc_meta !== undefined) {
        grpcMeta = {};
        // Normalize GRPC meta keys.
        for (const [key, value] of Object.entries(tomlProfile.grpc_meta)) {
            grpcMeta[(0, envconfig_toml_1.normalizeGrpcMetaKey)(key)] = value;
        }
    }
    const profile = {
        address: tomlProfile.address,
        namespace: tomlProfile.namespace,
        apiKey: tomlProfile.api_key,
        tls: fromTomlTLS(tomlProfile.tls),
        grpcMeta,
    };
    return (0, internal_workflow_1.filterNullAndUndefined)(profile);
}
/**
 * Converts a {@link ClientConfigProfile} to a TOML profile structure.
 *
 * @experimental Environment configuration is new feature and subject to change.
 */
function toTomlProfile(profile) {
    let grpc_meta = undefined;
    if (profile.grpcMeta !== undefined) {
        grpc_meta = {};
        // Normalize GRPC meta keys.
        for (const [key, value] of Object.entries(profile.grpcMeta)) {
            grpc_meta[(0, envconfig_toml_1.normalizeGrpcMetaKey)(key)] = value;
        }
    }
    const tomlProfile = {
        address: profile.address,
        namespace: profile.namespace,
        api_key: profile.apiKey,
        tls: toTomlTLS(profile.tls),
        grpc_meta,
    };
    return (0, internal_workflow_1.filterNullAndUndefined)(tomlProfile);
}
/**
 * Converts a TOML TLS configuration structure to a {@link ClientConfigTLS}.
 *
 * @experimental Environment configuration is new feature and subject to change.
 */
function fromTomlTLS(tomlTLS) {
    if (tomlTLS === undefined) {
        return undefined;
    }
    const clientConfigTLS = {
        disabled: tomlTLS.disabled,
        serverName: tomlTLS.server_name,
        clientCert: toConfigDataSource(tomlTLS.client_cert_path, tomlTLS.client_cert_data, 'client_cert'),
        clientKey: toConfigDataSource(tomlTLS.client_key_path, tomlTLS.client_key_data, 'client_key'),
        serverCACert: toConfigDataSource(tomlTLS.server_ca_cert_path, tomlTLS.server_ca_cert_data, 'server_ca_cert'),
    };
    return (0, internal_workflow_1.filterNullAndUndefined)(clientConfigTLS);
}
/**
 * Converts a {@link ClientConfigTLS} to a TOML TLS configuration structure.
 *
 * @experimental Environment configuration is new feature and subject to change.
 */
function toTomlTLS(tlsConfig) {
    if (tlsConfig === undefined) {
        return undefined;
    }
    const clientCert = toPathAndData(tlsConfig.clientCert);
    const clientKey = toPathAndData(tlsConfig.clientKey);
    const serverCACert = toPathAndData(tlsConfig.serverCACert);
    const tomlConfigTLS = {
        disabled: tlsConfig.disabled,
        server_name: tlsConfig.serverName,
        client_cert_path: clientCert?.path,
        client_cert_data: clientCert?.data && (0, encoding_1.decode)(clientCert?.data),
        client_key_path: clientKey?.path,
        client_key_data: clientKey?.data && (0, encoding_1.decode)(clientKey?.data),
        server_ca_cert_path: serverCACert?.path,
        server_ca_cert_data: serverCACert?.data && (0, encoding_1.decode)(serverCACert?.data),
    };
    return (0, internal_workflow_1.filterNullAndUndefined)(tomlConfigTLS);
}
/**
 * Converts a TOML client configuration structure to a {@link ClientConfig}.
 *
 * @experimental Environment configuration is new feature and subject to change.
 */
function fromTomlConfig(tomlConfig) {
    const profiles = {};
    for (const [profileName, profile] of Object.entries(tomlConfig.profile)) {
        profiles[profileName] = fromTomlProfile(profile);
    }
    return { profiles };
}
/**
 * Converts a {@link ClientConfig} to a TOML client configuration structure.
 *
 * @experimental Environment configuration is new feature and subject to change.
 */
function toTomlConfig(config) {
    const profile = {};
    for (const [profileName, configProfile] of Object.entries(config.profiles)) {
        profile[profileName] = toTomlProfile(configProfile);
    }
    return { profile };
}
function toPathAndData(source) {
    if (source === undefined) {
        return undefined;
    }
    if ('path' in source) {
        return { path: source.path };
    }
    if (typeof source.data === 'string') {
        return { data: (0, encoding_1.encode)(source.data) };
    }
    return { data: source.data };
}
function toConfigDataSource(path, data, fieldName) {
    if (path !== undefined && data !== undefined) {
        throw new Error(`Cannot specify both ${fieldName}_path and ${fieldName}_data`);
    }
    if (data !== undefined) {
        return { data: (0, encoding_1.encode)(data) };
    }
    if (path !== undefined) {
        return { path };
    }
    return undefined;
}
