// @temporalio/testing/lib/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimeSkippingWorkflowClient = exports.TimeSkippingClient = void 0;
require("abort-controller/polyfill"); // eslint-disable-line import/no-unassigned-import
const client_1 = require("@temporalio/client");
/**
 * A client with the exact same API as the "normal" client with one exception:
 * when `TestEnvClient.workflow` (an instance of {@link TimeSkippingWorkflowClient})
 * waits on a Workflow's result, it will enable time skipping in the Test Server.
 */
class TimeSkippingClient extends client_1.Client {
    constructor(options) {
        super(options);
        // Recreate the client (this isn't optimal but it's better than adding public methods just for testing).
        // NOTE: we cast to "any" to work around `workflow` being a readonly attribute.
        this.workflow = new TimeSkippingWorkflowClient({
            ...this.workflow.options,
            connection: options.connection,
        });
    }
}
exports.TimeSkippingClient = TimeSkippingClient;
/**
 * A client with the exact same API as the "normal" client with one exception: when this client
 * waits on a Workflow's result, it will enable time skipping in the Test Server.
 */
class TimeSkippingWorkflowClient extends client_1.WorkflowClient {
    testService;
    constructor(options) {
        super(options);
        const testService = options.connection.testService;
        if (!testService) {
            throw new TypeError('TestService must be present when creating a TimeSkippingWorkflowClient');
        }
        this.testService = testService;
    }
    /**
     * Gets the result of a Workflow execution.
     *
     * @see {@link WorkflowClient.result}
     */
    async result(workflowId, runId, opts) {
        await this.testService.unlockTimeSkipping({});
        try {
            return await super.result(workflowId, runId, opts);
        }
        finally {
            await this.testService.lockTimeSkipping({});
        }
    }
}
exports.TimeSkippingWorkflowClient = TimeSkippingWorkflowClient;
