// @temporalio/testing/lib/ephemeral-server.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toNativeEphemeralServerConfig = toNativeEphemeralServerConfig;
require("abort-controller/polyfill"); // eslint-disable-line import/no-unassigned-import
const time_1 = require("@temporalio/common/lib/time");
const pkg_1 = __importDefault(require("./pkg"));
/**
 * @internal
 */
function toNativeEphemeralServerConfig(server) {
    switch (server.type) {
        case 'dev-server':
            return {
                type: 'dev-server',
                exe: toNativeEphemeralServerExecutableConfig(server.executable),
                ip: server.ip ?? '127.0.0.1',
                port: server.port ?? null,
                ui: server.ui ?? false,
                uiPort: server.uiPort ?? null,
                namespace: server.namespace ?? 'default',
                dbFilename: server.dbFilename ?? null,
                log: server.log ?? { format: 'pretty', level: 'warn' },
                extraArgs: server.extraArgs ?? [],
            };
        case 'time-skipping':
            return {
                type: 'time-skipping',
                exe: toNativeEphemeralServerExecutableConfig(server.executable),
                port: server.port ?? null,
                extraArgs: server.extraArgs ?? [],
            };
        default:
            throw new TypeError(`Unsupported server type: ${String(server.type)}`);
    }
}
/**
 * @internal
 */
function toNativeEphemeralServerExecutableConfig(executable = { type: 'cached-download' }) {
    switch (executable.type) {
        case 'cached-download':
            return {
                type: 'cached-download',
                downloadDir: executable.downloadDir ?? null,
                version: executable.version ?? 'default',
                ttl: (0, time_1.msToNumber)(executable.ttl ?? '1d'),
                sdkName: 'sdk-typescript',
                sdkVersion: pkg_1.default.version,
            };
        case 'existing-path':
            return {
                type: 'existing-path',
                path: executable.path,
            };
        default:
            throw new TypeError(`Unsupported server executable type: ${String(executable.type)}`);
    }
}
