// @temporalio/testing/lib/mocking-activity-environment.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultActivityInfo = exports.MockActivityEnvironment = void 0;
require("abort-controller/polyfill"); // eslint-disable-line import/no-unassigned-import
const node_events_1 = __importDefault(require("node:events"));
const common_1 = require("@temporalio/common");
const logger_1 = require("@temporalio/common/lib/logger");
const worker_1 = require("@temporalio/worker");
const activity_1 = require("@temporalio/worker/lib/activity");
/**
 * An execution environment for testing Activities.
 *
 * Mocks Activity {@link Context | activity.Context} and exposes hooks for cancellation and heartbeats.
 *
 * Note that the `Context` object used by this environment will be reused for all activities that are run in this
 * environment. Consequently, once `cancel()` is called, any further activity that gets executed in this environment
 * will immediately be in a cancelled state.
 */
class MockActivityEnvironment extends node_events_1.default.EventEmitter {
    cancel = () => undefined;
    context;
    activity;
    constructor(info, opts) {
        super();
        const heartbeatCallback = (details) => this.emit('heartbeat', details);
        const loadedDataConverter = {
            payloadConverter: common_1.defaultPayloadConverter,
            payloadCodecs: [],
            failureConverter: common_1.defaultFailureConverter,
        };
        this.activity = new activity_1.Activity({ ...exports.defaultActivityInfo, ...info }, undefined, loadedDataConverter, heartbeatCallback, opts?.client, logger_1.LoggerWithComposedMetadata.compose(opts?.logger ?? new worker_1.DefaultLogger(), { sdkComponent: common_1.SdkComponent.worker }), opts?.metricMeter ?? common_1.noopMetricMeter, opts?.interceptors ?? []);
        this.context = this.activity.context;
        this.cancel = (reason, details) => {
            // Default to CANCELLED if nothing provided.
            const r = reason ?? 'CANCELLED';
            const d = details ?? new common_1.ActivityCancellationDetails({ cancelRequested: true });
            this.activity.cancel(r, d);
        };
    }
    /**
     * Run a function in Activity Context
     */
    async run(fn, ...args) {
        return this.activity.runNoEncoding(fn, { args, headers: {} });
    }
}
exports.MockActivityEnvironment = MockActivityEnvironment;
/**
 * Used as the default activity info for Activities executed in the {@link MockActivityEnvironment}
 *
 * @hidden
 */
exports.defaultActivityInfo = {
    attempt: 1,
    taskQueue: 'test',
    isLocal: false,
    taskToken: Buffer.from('test'),
    activityId: 'test',
    activityType: 'unknown',
    workflowType: 'test',
    base64TaskToken: Buffer.from('test').toString('base64'),
    heartbeatTimeoutMs: undefined,
    heartbeatDetails: undefined,
    activityNamespace: 'default',
    workflowNamespace: 'default',
    workflowExecution: { workflowId: 'test', runId: '00000000-0000-0000-0000-000000000000' },
    scheduledTimestampMs: 1,
    startToCloseTimeoutMs: 1000,
    scheduleToCloseTimeoutMs: 1000,
    currentAttemptScheduledTimestampMs: 1,
    priority: undefined,
    retryPolicy: undefined,
};
