// @temporalio/testing/lib/testing-workflow-environment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestWorkflowEnvironment = void 0;
require("abort-controller/polyfill"); // eslint-disable-line import/no-unassigned-import
const client_1 = require("@temporalio/client");
const connection_1 = require("@temporalio/client/lib/connection");
const common_1 = require("@temporalio/common");
const time_1 = require("@temporalio/common/lib/time");
const worker_1 = require("@temporalio/worker");
const core_bridge_1 = require("@temporalio/core-bridge");
const internal_workflow_1 = require("@temporalio/common/lib/internal-workflow");
const ephemeral_server_1 = require("./ephemeral-server");
const client_2 = require("./client");
/**
 * An execution environment for running Workflow integration tests.
 *
 * Runs an external server.
 * By default, the Java test server is used which supports time skipping.
 */
class TestWorkflowEnvironment {
    runtime;
    options;
    supportsTimeSkipping;
    server;
    address;
    /**
     * Namespace used in this environment (taken from {@link TestWorkflowEnvironmentOptions})
     */
    namespace;
    /**
     * Get an established {@link Connection} to the ephemeral server
     */
    connection;
    /**
     * A {@link TimeSkippingClient} for interacting with the ephemeral server
     */
    client;
    /**
     * An {@link AsyncCompletionClient} for interacting with the test server
     *
     * @deprecated - use `client.activity` instead
     */
    asyncCompletionClient;
    /**
     * A {@link TimeSkippingWorkflowClient} for interacting with the test server
     *
     * @deprecated - use `client.workflow` instead
     */
    workflowClient;
    /**
     * A {@link NativeConnection} for interacting with the test server.
     *
     * Use this connection when creating Workers for testing.
     */
    nativeConnection;
    constructor(runtime, options, supportsTimeSkipping, server, connection, nativeConnection, namespace, 
    /**
     * Address used when constructing `connection` and `nativeConnection`
     */
    address) {
        this.runtime = runtime;
        this.options = options;
        this.supportsTimeSkipping = supportsTimeSkipping;
        this.server = server;
        this.address = address;
        this.connection = connection;
        this.nativeConnection = nativeConnection;
        this.namespace = namespace;
        this.client = supportsTimeSkipping
            ? new client_2.TimeSkippingClient({
                connection,
                namespace: this.namespace,
                plugins: options.plugins,
                ...options.client,
            })
            : new client_1.Client({
                connection,
                namespace: this.namespace,
                plugins: options.plugins,
                ...options.client,
            });
        this.asyncCompletionClient = this.client.activity; // eslint-disable-line deprecation/deprecation
        this.workflowClient = this.client.workflow; // eslint-disable-line deprecation/deprecation
    }
    /**
     * Start a time skipping workflow environment.
     *
     * This environment automatically skips to the next events in time when a workflow handle's `result` is awaited on
     * (which includes {@link WorkflowClient.execute}). Before the result is awaited on, time can be manually skipped
     * forward using {@link sleep}. The currently known time can be obtained via {@link currentTimeMs}.
     *
     * This environment will be powered by the Temporal Time Skipping Test Server (part of the [Java SDK](https://github.com/temporalio/sdk-java)).
     * Note that the Time Skipping Test Server does not support full capabilities of the regular Temporal Server, and may
     * occasionally present different behaviors. For general Workflow testing, it is generally preferable to use {@link createLocal}
     * instead.
     *
     * Users can reuse this environment for testing multiple independent workflows, but not concurrently. Time skipping,
     * which is automatically done when awaiting a workflow result and manually done on sleep, is global to the
     * environment, not to the workflow under test. We highly recommend running tests serially when using a single
     * environment or creating a separate environment per test.
     *
     * By default, the latest release of the Test Server will be downloaded and cached to a temporary directory
     * (e.g. `$TMPDIR/temporal-test-server-sdk-typescript-*` or `%TEMP%/temporal-test-server-sdk-typescript-*.exe`). Note
     * that existing cached binaries will be reused without validation that they are still up-to-date, until the SDK
     * itself is updated. Alternatively, a specific version number of the Test Server may be provided, or the path to an
     * existing Test Server binary may be supplied; see {@link LocalTestWorkflowEnvironmentOptions.server.executable}.
     *
     * Note that the Test Server implementation may be changed to another one in the future. Therefore, there is no
     * guarantee that Test Server options, and particularly those provided through the `extraArgs` array, will continue to
     * be supported in the future.
     *
     * IMPORTANT: At this time, the Time Skipping Test Server is not supported on ARM platforms. Execution on Apple
     * silicon Macs will work if Rosetta 2 is installed.
     */
    static async createTimeSkipping(opts) {
        return await this.create({
            server: { type: 'time-skipping', ...opts?.server },
            client: opts?.client,
            plugins: opts?.plugins,
            supportsTimeSkipping: true,
        });
    }
    /**
     * Start a full Temporal server locally.
     *
     * This environment is good for testing full server capabilities, but does not support time skipping like
     * {@link createTimeSkipping} does. {@link supportsTimeSkipping} will always return `false` for this environment.
     * {@link sleep} will sleep the actual amount of time and {@link currentTimeMs} will return the current time.
     *
     * This local environment will be powered by [Temporal CLI](https://github.com/temporalio/cli), which is a
     * self-contained executable for Temporal. By default, Temporal's database will not be persisted to disk, and no UI
     * will be launched.
     *
     * By default, the latest release of the CLI will be downloaded and cached to a temporary directory
     * (e.g. `$TMPDIR/temporal-sdk-typescript-*` or `%TEMP%/temporal-sdk-typescript-*.exe`). Note that existing cached
     * binaries will be reused without validation that they are still up-to-date, until the SDK itself is updated.
     * Alternatively, a specific version number of the CLI may be provided, or the path to an existing CLI binary may be
     * supplied; see {@link LocalTestWorkflowEnvironmentOptions.server.executable}.
     *
     * Note that the Dev Server implementation may be changed to another one in the future. Therefore, there is no
     * guarantee that Dev Server options, and particularly those provided through the `extraArgs` array, will continue to
     * be supported in the future.
     */
    static async createLocal(opts) {
        return await this.create({
            server: { type: 'dev-server', ...opts?.server },
            client: opts?.client,
            plugins: opts?.plugins,
            namespace: opts?.server?.namespace,
            supportsTimeSkipping: false,
        });
    }
    /**
     * Create a new test environment using an existing server. You must already be running a server, which the test
     * environment will connect to.
     */
    static async createFromExistingServer(opts) {
        return await this.create({
            server: { type: 'existing' },
            client: opts?.client,
            plugins: opts?.plugins,
            namespace: opts?.namespace ?? 'default',
            supportsTimeSkipping: false,
            address: opts?.address,
        });
    }
    /**
     * Create a new test environment
     */
    static async create(opts) {
        const { supportsTimeSkipping, namespace, ...rest } = opts;
        const optsWithDefaults = addDefaults((0, internal_workflow_1.filterNullAndUndefined)(rest));
        let address;
        const runtime = worker_1.Runtime.instance();
        let server;
        if (optsWithDefaults.server.type !== 'existing') {
            // Add search attributes to CLI server arguments
            if ('searchAttributes' in optsWithDefaults.server && optsWithDefaults.server.searchAttributes) {
                let newArgs = [];
                for (const { name, type } of optsWithDefaults.server.searchAttributes) {
                    newArgs.push('--search-attribute');
                    newArgs.push(`${name}=${common_1.TypedSearchAttributes.toMetadataType(type)}`);
                }
                newArgs = newArgs.concat(optsWithDefaults.server.extraArgs ?? []);
                optsWithDefaults.server.extraArgs = newArgs;
            }
            server = await runtime.createEphemeralServer((0, ephemeral_server_1.toNativeEphemeralServerConfig)(optsWithDefaults.server));
            address = core_bridge_1.native.ephemeralServerGetTarget(server);
        }
        else {
            address = opts.address ?? 'localhost:7233';
            server = 'existing';
        }
        const nativeConnection = await worker_1.NativeConnection.connect({
            address,
            plugins: opts.plugins,
            [connection_1.InternalConnectionOptionsSymbol]: { supportsTestService: supportsTimeSkipping },
        });
        const connection = await client_1.Connection.connect({
            address,
            plugins: opts.plugins,
            [connection_1.InternalConnectionOptionsSymbol]: { supportsTestService: supportsTimeSkipping },
        });
        return new this(runtime, optsWithDefaults, supportsTimeSkipping, server, connection, nativeConnection, namespace, address);
    }
    /**
     * Kill the test server process and close the connection to it
     */
    async teardown() {
        await this.connection.close().catch((e) => {
            console.error(e);
            /* ignore */
        });
        await this.nativeConnection.close().catch((e) => {
            console.error(e);
            /* ignore */
        });
        if (this.server !== 'existing') {
            await this.runtime.shutdownEphemeralServer(this.server).catch((e) => {
                console.error(e);
                /* ignore */
            });
        }
    }
    /**
     * Wait for `durationMs` in "server time".
     *
     * This awaits using regular setTimeout in regular environments or manually skips time in time-skipping environments.
     *
     * Useful for simulating events far into the future like completion of long running activities.
     *
     * **Time skippping**:
     *
     * The time skippping server toggles between skipped time and normal time depending on what it needs to execute.
     *
     * This method is _likely_ to resolve in less than `durationMs` of "real time".
     *
     * @param durationMs number of milliseconds or {@link https://www.npmjs.com/package/ms | ms-formatted string}
     *
     * @example
     *
     * `workflow.ts`
     *
     * ```ts
     * const activities = proxyActivities({ startToCloseTimeout: 2_000_000 });
     *
     * export async function raceActivityAndTimer(): Promise<string> {
     *   return await Promise.race([
     *     wf.sleep(500_000).then(() => 'timer'),
     *     activities.longRunning().then(() => 'activity'),
     *   ]);
     * }
     * ```
     *
     * `test.ts`
     *
     * ```ts
     * const worker = await Worker.create({
     *   connection: testEnv.nativeConnection,
     *   activities: {
     *     async longRunning() {
     *       await testEnv.sleep(1_000_000); // <-- sleep called here
     *     },
     *   },
     *   // ...
     * });
     * ```
     */
    sleep = async (durationMs) => {
        if (this.supportsTimeSkipping) {
            await this.connection.testService.unlockTimeSkippingWithSleep({ duration: (0, time_1.msToTs)(durationMs) });
        }
        else {
            await new Promise((resolve) => setTimeout(resolve, (0, time_1.msToNumber)(durationMs)));
        }
    };
    /**
     * Get the current time known to this environment.
     *
     * For non-time-skipping environments this is simply the system time. For time-skipping environments this is whatever
     * time has been skipped to.
     */
    async currentTimeMs() {
        if (this.supportsTimeSkipping) {
            const { time } = await this.connection.testService.getCurrentTime({});
            return (0, time_1.tsToMs)(time);
        }
        else {
            return Date.now();
        }
    }
}
exports.TestWorkflowEnvironment = TestWorkflowEnvironment;
function addDefaults(opts) {
    return {
        client: {},
        ...opts,
        server: {
            ...opts.server,
        },
        plugins: [],
    };
}
