// @temporalio/ai-sdk/lib/mcp.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TemporalMCPClient = void 0;
const ai_1 = require("ai");
const workflow = __importStar(require("@temporalio/workflow"));
/**
 * A Temporal MCP Client which uses activities to execute list tools and call tools.
 * This should match by name an MCPClientFactory registered with the AI SDK plugin to function.
 *
 * This is intentionally similar to the AI SDK's ``MCPClient``, but is lacking functions not used by the framework.
 * It doesn't close either since the underlying clients are created in activities.
 *
 * @experimental The AI SDK plugin is an experimental feature; APIs may change without notice.
 */
class TemporalMCPClient {
    options;
    constructor(options) {
        this.options = options;
    }
    async tools() {
        const activities = workflow.proxyActivities({
            startToCloseTimeout: '10 minutes',
            ...this.options.activityOptions,
        });
        const listActivity = activities[this.options.name + '-listTools'];
        const tools = await listActivity({ clientArgs: this.options.clientArgs });
        return Object.fromEntries(Object.entries(tools).map(([toolName, toolResult]) => [
            toolName,
            {
                execute: async (input, options) => {
                    const activities = workflow.proxyActivities({
                        summary: toolName,
                        startToCloseTimeout: '10 minutes',
                        ...this.options,
                    });
                    const callActivity = activities[this.options.name + '-callTool'];
                    return await callActivity({ name: toolName, input, options, clientArgs: this.options.clientArgs });
                },
                inputSchema: (0, ai_1.jsonSchema)(toolResult.inputSchema),
                type: 'dynamic',
            },
        ]));
    }
}
exports.TemporalMCPClient = TemporalMCPClient;
