// @temporalio/ai-sdk/lib/provider.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.temporalProvider = exports.TemporalProvider = exports.TemporalEmbeddingModel = exports.TemporalLanguageModel = void 0;
const workflow = __importStar(require("@temporalio/workflow"));
const common_1 = require("@temporalio/common");
/**
 * A language model implementation that delegates AI model calls to Temporal activities.
 * This allows workflows to invoke AI models through the Temporal execution model.
 *
 * @experimental The AI SDK integration is an experimental feature; APIs may change without notice.
 */
class TemporalLanguageModel {
    modelId;
    options;
    specificationVersion = 'v3';
    provider = 'temporal';
    constructor(modelId, options) {
        this.modelId = modelId;
        this.options = options;
    }
    get supportedUrls() {
        return {};
    }
    async doGenerate(options) {
        const activities = workflow.proxyActivities({
            startToCloseTimeout: '10 minutes',
            ...this.options,
        });
        const result = await activities.invokeModel({ modelId: this.modelId, options });
        if (result === undefined) {
            throw common_1.ApplicationFailure.nonRetryable('Received undefined response from model activity.');
        }
        if (result.response !== undefined && result.response.timestamp) {
            const timestamp = new Date(result.response.timestamp);
            // Only set if it's a valid date
            if (!isNaN(timestamp.getTime())) {
                result.response.timestamp = timestamp;
            }
        }
        return result;
    }
    doStream(_options) {
        throw common_1.ApplicationFailure.nonRetryable('Streaming not supported.');
    }
}
exports.TemporalLanguageModel = TemporalLanguageModel;
/**
 * An embedding model implementation that delegates embedding generation to Temporal activities.
 * This allows workflows to generate embeddings through the Temporal execution model.
 *
 * @experimental The AI SDK integration is an experimental feature; APIs may change without notice.
 */
class TemporalEmbeddingModel {
    modelId;
    options;
    specificationVersion = 'v3';
    provider = 'temporal';
    /**
     * Undefined to let the underlying provider handle chunking, as it knows its own limits.
     */
    maxEmbeddingsPerCall = undefined;
    /**
     * Indicates the underlying embedding model API can handle concurrent requests.
     * Set to true since we delegate to the actual provider which manages its own concurrency.
     */
    supportsParallelCalls = true;
    constructor(modelId, options) {
        this.modelId = modelId;
        this.options = options;
    }
    async doEmbed(options) {
        const activities = workflow.proxyActivities({
            startToCloseTimeout: '10 minutes',
            ...this.options,
        });
        const result = await activities.invokeEmbeddingModel({
            modelId: this.modelId,
            values: options.values,
            providerOptions: options.providerOptions,
            headers: options.headers,
            // Note: abortSignal is not serializable, Temporal's cancellation handles this
        });
        if (result === undefined) {
            throw common_1.ApplicationFailure.nonRetryable('Received undefined response from embedding model activity.');
        }
        return result;
    }
}
exports.TemporalEmbeddingModel = TemporalEmbeddingModel;
/**
 * A Temporal-specific provider implementation that creates AI models which execute
 * through Temporal activities. This provider integrates AI SDK models with Temporal's
 * execution model to ensure reliable, durable AI model invocations.
 *
 * @experimental The AI SDK integration is an experimental feature; APIs may change without notice.
 */
class TemporalProvider {
    options;
    specificationVersion = 'v3';
    constructor(options) {
        this.options = options;
    }
    imageModel(_modelId) {
        throw new Error('Not implemented');
    }
    languageModel(modelId) {
        return new TemporalLanguageModel(modelId, {
            ...this.options?.default,
            ...this.options?.languageModel,
        });
    }
    embeddingModel(modelId) {
        return new TemporalEmbeddingModel(modelId, {
            ...this.options?.default,
            ...this.options?.embeddingModel,
        });
    }
    transcriptionModel(_modelId) {
        throw new Error('Not implemented');
    }
}
exports.TemporalProvider = TemporalProvider;
/**
 * A singleton instance of TemporalProvider for convenient use in applications.
 *
 * @experimental The AI SDK integration is an experimental feature; APIs may change without notice.
 */
exports.temporalProvider = new TemporalProvider();
