// @temporalio/ai-sdk/lib/provider.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.temporalProvider = exports.TemporalProvider = exports.TemporalLanguageModel = void 0;
const workflow = __importStar(require("@temporalio/workflow"));
const common_1 = require("@temporalio/common");
/**
 * A language model implementation that delegates AI model calls to Temporal activities.
 * This allows workflows to invoke AI models through the Temporal execution model.
 *
 * @experimental The AI SDK integration is an experimental feature; APIs may change without notice.
 */
class TemporalLanguageModel {
    modelId;
    options;
    specificationVersion = 'v2';
    provider = 'temporal';
    supportedUrls = {};
    constructor(modelId, options) {
        this.modelId = modelId;
        this.options = options;
    }
    async doGenerate(options) {
        const activities = workflow.proxyActivities({ startToCloseTimeout: '10 minutes', ...this.options });
        const result = await activities.invokeModel({ modelId: this.modelId, options });
        if (result === undefined) {
            throw common_1.ApplicationFailure.nonRetryable('Received undefined response from model activity.');
        }
        if (result.response !== undefined) {
            result.response.timestamp = new Date(result.response.timestamp);
        }
        return result;
    }
    doStream(_options) {
        throw common_1.ApplicationFailure.nonRetryable('Streaming not supported.');
    }
}
exports.TemporalLanguageModel = TemporalLanguageModel;
/**
 * A Temporal-specific provider implementation that creates AI models which execute
 * through Temporal activities. This provider integrates AI SDK models with Temporal's
 * execution model to ensure reliable, durable AI model invocations.
 *
 * @experimental The AI SDK integration is an experimental feature; APIs may change without notice.
 */
class TemporalProvider {
    options;
    constructor(options) {
        this.options = options;
    }
    imageModel(_modelId) {
        throw new Error('Not implemented');
    }
    languageModel(modelId) {
        return new TemporalLanguageModel(modelId, this.options);
    }
    textEmbeddingModel(_modelId) {
        throw new Error('Not implemented');
    }
}
exports.TemporalProvider = TemporalProvider;
/**
 * A singleton instance of TemporalProvider for convenient use in applications.
 *
 * @experimental The AI SDK integration is an experimental feature; APIs may change without notice.
 */
exports.temporalProvider = new TemporalProvider();
